docs/recipes/ui/add-routing-with-separate-states.md

```diff
@@ -0,0 +1,292 @@
+
+```fsharp title="Index.fs"
+let initFromUrl url =
+    match url with
+    | [ "todo" ] ->
+        let todoListModel, todoListMsg = TodoList.init ()
+        let model = { CurrentPage = TodoList todoListModel }
+
+        model, todoListMsg |> Cmd.map TodolistMsg
+```
+
+Add a wildcard, so any URLs that are not registered display the "not found" page
+
+=== "Code"
+    ```fsharp title="Index.fs"
+    let initFromUrl url =
+        match url with
+        ...
+        | _ -> { CurrentPage = NotFound }, Cmd.none
+    ```
+=== "Diff"
+    ```.diff title="Index.fs"
+     let initFromUrl url =
+         match url with
+         ...
+    +    | _ -> { CurrentPage = NotFound }, Cmd.none
+    ```
+
+## 6. Elmish initialization
+
+Add an `init` function to `Index`; return the current page based on `Router.currentPath`
+
+```fsharp title="Index.fs"
+let init () : Model * Cmd<Msg> =
+    Router.currentUrl ()
+    |> initFromUrl
+```
+
+## 7. Handling URL Changes
+
+Add an `UrlChanged` case of `string list` to the `Msg` type
+
+=== "Code"
+    ```fsharp title="Index.fs"
+    type Msg =
+        ...
+        | UrlChanged of string list
+    ```
+=== "Diff"
+    ```diff title="Index.fs"
+     type Msg =
+         ...
+    +    | UrlChanged of string list
+    ```
+
+Handle the case in the `update` function by calling `initFromUrl`
+
+=== "Code"
+    ```fsharp title="Index.fs"
+    let update (message: Msg) (model: Model) : Model * Cmd<Msg> =
+        ...
+        match model.CurrentPage, message with
+        | _, UrlChanged url -> initFromUrl url
+    ```
+=== "Diff"
+    ```diff title="Index.fs"
+     let update (message: Msg) (model: Model) : Model * Cmd<Msg> =
+         ...
+    +    match model.CurrentPage, message with
+    +    | _, UrlChanged url -> initFromUrl url
+    ```
+
+## 8. Catching all cases in the update function
+
+Complete the pattern match in the `update` function, adding a case with a wildcard for both `message` and `model`. Return the model, and no command
+
+=== "Code"
+    ```fsharp title="Index.fs"
+    let update (message: Msg) (model: Model) : Model * Cmd<Msg> =
+        ...
+        | _, _ -> model, Cmd.none
+    ```
+=== "Diff"
+    ```.diff title="Index.fs"
+     let update (message: Msg) (model: Model) : Model * Cmd<Msg> =
+         ...
+    +    | _, _ -> model, Cmd.none
+    ```
+## 9. Rendering pages
+
+Add a function containerBox to the `Index` module. If the CurrentPage is of `TodoList`, render the todo list using `TodoList.view`; in order to dispatch a `TodoList.Msg`, it needs to be wrapped in a `TodoListMsg`.
+
+For the `NotFound` page, return a "Page not found" box
+
+```fsharp title="Index.fs"
+let containerBox model dispatch =
+    match model.CurrentPage with
+    | TodoList todoModel -> TodoList.view todoModel ( TodolistMsg>>dispatch )
```

Comment:
It's conventional to have spaces around operators and not inside parentheses

```suggestion
    | TodoList todoModel -> TodoList.view todoModel (TodolistMsg >> dispatch)
```

You might find the official [F# code formatting guidelines](https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting) useful to refer to occasionally.